[3단계 - 질문 삭제하기 리팩터링] 리뷰요청드립니다~

byunghakjang1230 opened the pull request:
리뷰요청 드립니다!
처음에는 너무 막연 했었는데 리팩토링 미션 소개 동영상을 보고 나니 감이 조금 잡혔습니다. 그래서 그런지 생각보다 큰 어려움 없이 진행은 했는데 괜찮은 방향으로 한건지는 잘 모르겠네요.^^;; 마지막 단계인 만큼 뼈때리는 리뷰 부탁드립니다!

hilu0318 committed:
docs(README.md): 문서 내용 정리 및 다른 문서파일로 이관

docs(STEP1_SUMMARY.md): step1 미션에 대한 요구사항 및 학습내용 정리

docs(STEP3_SUMMARY.md): step3 미션에 대한 요구사항 정리

hilu0318 committed:
docs(STEP3_SUMMARY.md): 요구사항 구현 확인 내용 수정

feat(Answer.java): 로그인 사용자의 정보를 받아 삭제 상태값 수정하는 기능 추가

- 입력받은 사용자와 작성자 비교 후 예외처리
- 삭제 상태값 수정이 완료되면 삭제이력을 반환

test(AnswerTest.java): 삭제처리 기능 테스트 코드 작성

- 입력받은 사용자와 작성자 비교하여 다를경우 예외상황 발생 테스트
- 입력받은 사용자와 작성자가 같을 경우 삭제 상태값 수정 테스트
- 삭제 상태값 수정 완료 후 삭제이력 반환 테스트

feat(Question.java): 로그인 사용자의 정보를 받아 삭제 상태값 수정하는 기능 추가

- 입력받은 사용자와 작성자 비교 후 예외처리
- 삭제 상태값 수정이 완료되면 삭제이력을 반환

test(QuestionTest.java): 삭제처리 기능 테스트 코드 작성

- 입력받은 사용자와 작성자 비교하여 다를경우 예외상황 발생 테스트
- 입력받은 사용자와 작성자가 같을 경우 삭제 상태값 수정 테스트
- 삭제 상태값 수정 완료 후 삭제이력 반환 테스트

hilu0318 committed:
docs(STEP3_SUMMARY.md): 요구사항 구현 확인 내용 수정

feat(Answers.java): Answer 리스트를 관리하는 일급 컬렉션 구현

- 사용자 정보를 입력받아 답변 목록 일괄 삭제상태 변경 처리하는 기능
- 삭제 상태값 수정이 완료되면 삭제이력 목록을 반환

test(AnswersTest.java): Answers 클래스에 구현된 기능 검증

- 입력받은 사용자와 답변 작성자들과 비교하여 다른 경우가 있으면 예외상황 발생 테스트
- 모든 목록의 삭제상태값 수정 완료 시 삭제이력 목록 반환

refactor(DeleteHistory.java): getter 메소드 추가

hilu0318 committed:
docs(STEP3_SUMMARY.md): 요구사항 구현 확인 내용 수정

refactor(QnaService.java): 게시물 삭제처리 로직을 도메인으로 옮긴 뒤 리팩터링 진행

- QnaServiceTest 전체 테스트 코드 통과

hilu0318 committed:
docs(STEP3_SUMMARY.md): 요구사항 내용 수정

hilu0318 committed:
refactor(QnaService.java): 메소드 내 변수 수정 등 리팩터링 진행

young891221 commented on src/main/java/qna/service/QnaService.java:
```diff
@@ -40,25 +39,13 @@ public Question findQuestionById(Long id) {
-            }
-        }
+        Answers answers = new Answers(answerRepository.findByQuestionIdAndDeletedFalse(question.getId()));
+        deleteHistories.addAll(answers.deleteAllByWriter(loginUser));
```
`DeleteHistories`라는 객체를 만들어서 모두 묶어주면 어떨까요?ㅎ

byunghakjang1230 replied:
일급 컬렉션 만들어서 포장하였습니다. 매개변수를 어떻게 받아서 초기화 해야 생성자에서 로직이 최소화될지 고민을 했었는데..비슷한 모양으로 나올거 같아서 Question, Answers, User를 받아서 초기화하도록 했습니다. 생성자 초기화 매개변수로 3개를 넘기는게 좀 신경쓰이긴 하네요...😓 서비스 영역에 코드가 조금 늘어나더라도 생성자 수를 줄이는 방향이 좋은걸까요? 🤔

young891221 replied:
음 개인적으로는 생성자에 늘어나도 위임시키는게 좋아 보입니다ㅎ

byunghakjang1230 replied:
답변 감사합니다~ 참고하도록 하겠습니다!😀

young891221 commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -104,6 +102,14 @@ public void setDeleted(boolean deleted) {
     }
 
+    public DeleteHistory deleteByWriter(User loginUser) throws CannotDeleteException {
+        if (!this.isOwner(loginUser)) {
+            throw new CannotDeleteException("질문을 삭제할 권한이 없습니다.");
+        }
```
validation 메서드 분리시켜 주세요!

byunghakjang1230 replied:
넵, 별도의 메소드로 분리시켰습니다~!

young891221 reviewed (changes requested):
기능부터 테스트까지 잘 구현되어 있네요 💯 
몇 가지 피드백 드렸어요. 확인 부탁드립니다!

hilu0318 committed:
refactor(Question.java): 유효성 검증 메소드 분리

refactor(Answer.java): 유효성 검증 메소드 분리

hilu0318 committed:
feat(DeleteHistories.java): 삭제 이력 목록을 포장하는 일급 컬렉션 구현

- Question, Answers, User를 받아 삭제이력목록 초기화 작업

test(DeleteHistoriesTest.java): DeleteHistories 객체 생성자 및 기능 테스트

refactor(QnaService.java): DeleteHistories 클래스를 사용한 리팩토링 진행

young891221 reviewed (approved):
리뷰 반영 잘 해주셨네요 💯 
JPA 미션하느라 고생하셨습니다. 머지할게요!
